### PR TITLE
Added RefreshToken method

### PR DESCRIPTION
**Pull request by snobu**
* Added `RefreshToken` method to TokenCache.ts

Not yet used by our code, but ready to be used.

## What's this for and how do we use it?

Instead of calling the HLS URL with the access token we extract from sessionInfo (on login), we take the same access token, call `TokenCache.RefreshToken()` and get back the value of a `Cookie` header that we can now pass to ffmpeg. Unfortunately the fresh token is now a cookie, i am unable to get a fresh Bearer token. 

This cookie, if unrolled as an opaque access token inside, but that doesn't stand on its own as Bearer (returns 401 Unauthorized) so the full cookie must be sent as `Cookie:`.

MS Stream API works in mysterious ways.

## Why?

This addresses https://github.com/snobu/destreamer/issues/43, each video in _to be downloaded_ array can now be downloaded with a fresh token that's good for one full hour. Since the validation of the token is performed remotely only once (when we send the request), even if we take two hours to download a video it won't matter, it will succeed.

We should call RefreshToken even for the first video in our array since the cached access token may only have a few minutes left before expiring.

The metadata is fetched for all videos on startup (with the original access token), so we don't need to worry about that part.

## Caveats?

ffmpeg must now be called with `Cookie: <value returned by RefreshToken()>` instead of an Authorization header.

`RefreshToken()` may come back as null (as the API call that i make is highly experimental), so our implementation can decide if we continue with a new "cookie" token or we fall back to our original access token.

## Sample usage

```javascript
let freshCookie: string | null = tokenCache.RefreshToken(session);
if (freshCookie) {
    // Call ffmpeg with a `Cookie: <value of freshCookie>` header
}
else {
    // Call ffmpeg with original access token (Authorization: Bearer ...)
}
```

We should have a method that returns the corrent header for ffmpeg so we don't pollute the already complex ffmpeg wrapper with unnecessary ifs.

### CODE REVIEW

**Comment by lukaarma** [comment]
Seems good to me and i think that we can keep calling the api with the cookie we get back in case we have really long downloads, like keep asking for a new cookie after each download

**Comment by snobu** [comment]
Yes, that's the plan, before each new download call RefreshToken(). We should call this even for the first video in the array since the access token in cache may only have a few minutes left before expiration.

File: src/TokenCache.ts [hunk 1]
```diff
@@ -53,4 +54,27 @@ export class TokenCache {
...
+            });
+
+        let freshCookie: string | null = null;
+        
+        try {
+            let cookie: string = response.headers["set-cookie"].toString();
+            let freshCookie = cookie.split(',Authorization_Api=')[0];
+        }
+        catch (e) {
+            console.error("Error when calling /refreshtoken: Missing or unexpected set-cookie header.");
```
Comment (kylon) [inline src/TokenCache.ts]: maybe we should process.exit here since the next download will not succeed?

**Comment by kylon** [comment]
we can do something like this
(fussonia)

```
fn () {
logic...
return ['header/cookie', 'value'];
}

donwloadvideo() {
....

let token;

for...
token = fn()
const args = [something]
args.concat(token)
spawn ffmpeg with args
endfor

...
}
```

**Comment by snobu** [comment]
Closing. The fix is now in master.

**Pull request closed without merging**